File: src/models/effect/effect.py
import enum
from sqlalchemy import Column, Integer, Enum
from src.db.base import Base


class EffectType(int, enum.Enum):
    PUSH = 1
    PULL = 2
    FORCED_MOVEMENT_IMMUNITY = 3
    POISON = 4
    POISON_IMMUNITY = 5
    FIRE = 6
    FIRE_IMMUNITY = 7
    BLEED = 8
    BLEED_IMMUNITY = 9
    DISARM = 10
    DISARM_IMMUNITY = 11
    STUN = 12
    STUN_IMMUNITY = 13
    CONFUSION = 14
    CONFUSION_IMMUNITY = 15
    CHARM = 16
    CHARM_IMMUNITY = 17
    FEAR = 18
    FEAR_IMMUNITY = 19
    INVISIBILITY = 20
    SHIELD = 21
    BONUS_HEALTH = 22
    BONUS_DAMAGE = 23
    BONUS_MOVEMENT = 24

    def __repr__(self):
        return self.name


class EffectRange(enum.Enum):
    SELF = 1
    ONE = 2
    ALL = 3

    def __repr__(self):
        return self.name


class Effect(Base):
    """
    Effect model

    "id"        INT [not null, increment]
    "type"      EffectType [not null]
    "duration"  INT [not null]
    "range"     EffectRange [not null]
    "strength"  INT
    """
    id = Column(Integer, primary_key=True)
    type = Column(Enum(EffectType), nullable=False)
    duration = Column(Integer, nullable=False)
    range = Column(Enum(EffectRange), nullable=False)
    strength = Column(Integer, nullable=True)

    def __repr__(self):
        return f"<Effect(id={self.id}, type={self.type}, duration={self.duration}, range={self.range}, strength={self.strength})>"

    def __str__(self):
        return (f"Effect "
                f"  id={self.id}, "
                f"  type={self.type.value}, "
                f"  duration={self.duration}, "
                f"  range={self.range.value}, "
                f"  strength={self.strength} ")

File: src/main/java/cz/trailsthroughshadows/api/table/effect/Effect.java
package cz.trailsthroughshadows.api.table.effect;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;

import java.io.Serializable;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "Effect")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Effect {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Setter(AccessLevel.NONE)
    private Integer id;

    @Column
    @Enumerated(EnumType.STRING)
    private EffectTarget target;

    @Column
    @Enumerated(EnumType.STRING)
    private EffectType type;

    @Column
    private int duration;

    @Column
    private Integer strength;

    public enum EffectType implements Serializable {
        PUSH,
        PULL,
        FORCED_MOVEMENT_RESISTANCE,
        POISON,
        POISON_RESISTANCE,
        FIRE,
        FIRE_RESISTANCE,
        BLEED,
        BLEED_RESISTANCE,
        DISARM,
        DISARM_RESISTANCE,
        STUN,
        STUN_RESISTANCE,
        CONFUSION,
        CONFUSION_RESISTANCE,
        CHARM,
        CHARM_RESISTANCE,
        FEAR,
        FEAR_RESISTANCE,
        INVISIBILITY,
        SHIELD,
        HEAL,
        REGENERATION,
        BONUS_HEALTH,
        BONUS_DAMAGE,
        BONUS_MOVEMENT,
    }

    public enum EffectTarget implements Serializable {
        SELF,
        ONE,
        ALL,
        ALL_ALLIES,
        ALL_ENEMIES,
    }
}
